SHIP-0026 Shipwright-managed push (#1046)
shipwright-io/build: docs/buildstrategies.md

| ------------------------------ | ----------- |
| `$(params.shp-source-root)`    | The absolute path to the directory that contains the user's sources. |
| `$(params.shp-source-context)` | The absolute path to the context directory of the user's sources. If the user specified no value for `spec.source.contextDir` in their `Build`, then this value will equal the value for `$(params.shp-source-root)`. Note that this directory is not guaranteed to exist at the time the container for your step is started, you can therefore not use this parameter as a step's working directory. |
| `$(params.shp-output-directory)` | The absolute path to a directory that the build strategy should store the image in. You can store a single tarball containing a single image, or an OCI image layout. |

> | `$(params.shp-output-image)`     | The URL of the image that the user wants to push as specified in the Build's `spec.output.image`, or the override from the BuildRun's `spec.output.image`. |

Reviewer: ```suggestion
| `$(params.shp-output-image)`     | The URL of the image that the user wants to push, as specified in the Build's `spec.output.image` or as an override from the BuildRun's `spec.output.image`. |
```

> | `$(parama.shp-output-insecure)`  | Flag that indicates if the output image's registry location is insecure, which means that it uses a certificated not signed by a certificate authority, or uses HTTP |

Reviewer: ```suggestion
| `$(parama.shp-output-insecure)`  | A flag that indicates the output image's registry location is insecure because it uses a certificate not signed by a certificate authority, or uses HTTP |
```

> ### Output directory vs output image

Reviewer: ```suggestion
### Output directory vs. output image
```

> As a build strategy author, you decide whether you just build the image, or also push it to the container registry. You implicitly indicate how your strategy behaves by using, or not using the `$(params.shp-output-image)` system parameter:

Reviewer: ```suggestion
As a build strategy author, you decide whether to build the image or push it to the container registry. You implicitly indicate how your strategy behaves by using or not using the `$(params.shp-output-image)` system parameter:
```

Author: This is not correct. The build strategy always must build a container image. It can also push the image.

- If you DO NOT use `$(params.shp-output-image)`, then Shipwright assumes that your build strategy PUSHES the image. We call this a strategy-managed push.
- If you DO use `$(params.shp-output-image)`, then Shipwright assumes that your build strategy does NOT PUSH the image. We call this a shipwright-managed push.

> When you use the `$(params.shp-output-image)` parameter, then Shipwright will also take care to set the [image-related system results](#system-results).

Reviewer: ```suggestion
When you use the `$(params.shp-output-image)` parameter, then Shipwright also takes care to set the [image-related system results](#system-results).
```
Can we use a more specific phrase than "takes care to set"? Does that mean Shipwright sets the system results, or it ensures that someone performs an action?

Author: Thanks. Changed it to

> When you use the `$(params.shp-output-image)` parameter, then Shipwright will also set the [image-related system results](#system-results).

If you are uncertain about how to implement your build strategy, then follow this guidance:

> 1. If your build strategy tool is unable to locally store an image but always pushes it, then you must do the push operation. An example is the [Buildpacks strategy](../samples/buildstrategy/buildpacks-v3/buildstrategy_buildpacks-v3_cr.yaml). You SHOULD respect the `$(params.shp-output-insecure)` parameter.

Reviewer: ```suggestion
1. If your build strategy tool cannot locally store an image but always pushes it, then you must do the push operation. An example is the [Buildpacks strategy](../samples/buildstrategy/buildpacks-v3/buildstrategy_buildpacks-v3_cr.yaml). You SHOULD respect the `$(params.shp-output-insecure)` parameter.
```

2. If your build strategy tool can locally store an image, then the choice depends on how you expect your build users to make use of your strategy, and the nature of your strategy.

>    1. Some build strategies do not produce all layers of an image, but only a layer or some layers with the application, and use a common base image. An example is `ko`. Such base image layers are often already present in the destination registry (like in rebuilds). If the strategy can perform the push operation, then it can optimize the process and can omit the download of the base image when it is not required to push it. In case of a shipwright-manager push, the complete image must be locally store in `$(params.shp-output-directory)` which implies that a base image must always be downloaded.

Reviewer: ```suggestion
   1. Some build strategies do not produce all layers of an image, but only a layer or some layers within the application, and use a common base image. An example is `ko`. Such base image layers are often already present in the destination registry (like in rebuilds). If the strategy can perform the push operation, then it can optimize the process and can omit the download of the base image when it is not required to push it. In the case of a shipwright-manager push, the complete image must be locally stored in `$(params.shp-output-directory)`, which implies that a base image must always be downloaded.
```

Author: "within" is inaccurate, but I get your point. I rephrased it.

>    2. Some build strategy tools do not make it easy to determine the digest or size of the image. This can make it complex to set the [strategy results](#system-results). In case of a shipwright-managed push, Shipwright has the responsibility to set them.

Reviewer: ```suggestion
   2. Some build strategy tools do not make it easy to determine the digest or size of the image, which can make it complex for you to set the [strategy results](#system-results). In the case of a shipwright-managed push, Shipwright has the responsibility to set them.
```

>    3. Build users can configure the build to amend additional annotations, or labels to the final image. In case of a shipwright-managed push, these can be set directly and the image will only be pushed once. In a strategy-managed push scenario, your build strategy will push a first version of the image without those annotations and labels. Shipwright will then mutate the image and push it again with the updated annotations and labels. Such a duplicate push can cause unexpected behavior with registries that trigger other actions when an image gets pushed, or that do not allow overwriting a tag.

Reviewer: ```suggestion
   3. Build users can configure the build to amend additional annotations or labels to the final image. In the case of a shipwright-managed push, these can be set directly and the image will only be pushed once. In a strategy-managed push scenario, your build strategy will push the first version of the image without those annotations and labels. Shipwright will then mutate the image and push it again with the updated annotations and labels. Such a duplicate push can cause unexpected behavior with registries that trigger other actions when an image gets pushed, or that do not allow overwriting a tag.
```

| ------------------------------ | ----------- |
| `$(params.shp-source-root)`    | The absolute path to the directory that contains the user's sources. |
| `$(params.shp-source-context)` | The absolute path to the context directory of the user's sources. If the user specified no value for `spec.source.contextDir` in their `Build`, then this value will equal the value for `$(params.shp-source-root)`. Note that this directory is not guaranteed to exist at the time the container for your step is started, you can therefore not use this parameter as a step's working directory. |
| `$(params.shp-output-directory)` | The absolute path to a directory that the build strategy should store the image in. You can store a single tarball containing a single image, or an OCI image layout. |
| `$(params.shp-output-image)`     | The URL of the image that the user wants to push, as specified in the Build's `spec.output.image` or as an override from the BuildRun's `spec.output.image`. |
| `$(params.shp-output-insecure)`  |  A flag that indicates the output image's registry location is insecure because it uses a certificate not signed by a certificate authority, or uses HTTP. |

> ### Output directory vs. output image

Reviewer 2: this explanation is great, kudos !

> As a build strategy author, you decide whether you just build the image, or also push it to the container registry. You implicitly indicate how your strategy behaves by using, or not using the `$(params.shp-output-image)` system parameter:

Reviewer: Though logical, the current wording misdirects some of the reader's attention away from the key decision. The following suggestion highlights the key decision.
```suggestion
As a build strategy author, you decide whether your build strategy or Shipwright pushes the build image to the container registry. You do this by using or omitting the `$(params.shp-output-image)` system parameter:
```

Author: Done.
